File: optimization-heap-operations-problems-enhanced/src/main/java/com/codinginterview/heapoperations/problems/TopKFrequentElements.java
```java
package com.codinginterview.heapoperations.problems;

import com.codinginterview.heapoperations.utils.CustomMinHeap;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

/**
 * Problem: Top K Frequent Elements
 *
 * Given an integer array nums and an integer k, return the k most frequent elements.
 * You may return the answer in any order.
 *
 * Example 1:
 * Input: nums = [1,1,1,2,2,3], k = 2
 * Output: [1,2] (or [2,1])
 *
 * Example 2:
 * Input: nums = [1], k = 1
 * Output: [1]
 *
 * Constraints:
 * 1 <= nums.length <= 10^5
 * -10^4 <= nums[i] <= 10^4
 * k is in the range [1, the number of unique elements in the array].
 * It is guaranteed that the answer is unique, meaning there is only one set of the k most frequent elements.
 */
public class TopKFrequentElements {

    /**
     * Approach 1: Optimal Solution using HashMap + Min-Heap (Java's PriorityQueue)
     *
     * Algorithm:
     * 1. **Count Frequencies:** Use a HashMap to store the frequency of each number.
     *    Iterate through the input array `nums` and populate `Map<Integer, Integer>`
     *    where key is the number and value is its frequency.
     *    Time: O(N), Space: O(M) where M is the number of unique elements.
     *
     * 2. **Maintain Top K with Min-Heap:** Create a Min-Heap (PriorityQueue) that
     *    stores `Map.Entry<Integer, Integer>` objects. The heap should be ordered
     *    by the *frequency* (value of the entry) in ascending order.
     *    Iterate through each entry in the frequency map:
     *    a. Add the entry to the min-heap.
     *    b. If the size of the min-heap exceeds `k`, remove the entry with the smallest
     *       frequency (`heap.poll()`). This ensures the heap always contains `k` elements
     *       that have the highest frequencies encountered so far. The root of this heap
     *       will always be the element with the Kth highest frequency.
     *    Time: O(M log K), Space: O(K)
     *
     * 3. **Extract Results:** Create an `int[]` array of size `k`.
     *    While the heap is not empty, extract elements (`heap.poll()`) and add their keys
     *    (the actual numbers) to the result array.
     *    Time: O(K log K) (or O(K) if simply moving to list/array), Space: O(K)
     *
     * Overall Time Complexity: O(N + M log K)
     *   - N for frequency counting, M for iterating unique elements for heap, K for final extraction.
     * Overall Space Complexity: O(M + K)
     *   - M for HashMap, K for Min-Heap.
     *
     * @param nums The input integer array.
     * @param k The number of most frequent elements to return.
     * @return An array containing the k most frequent elements.
     */
    public int[] topKFrequentHeap(int[] nums, int k) {
        // Step 1: Count frequencies of each number
        Map<Integer, Integer> freqMap = new HashMap<>();
        for (int num : nums) {
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }

        // Step 2: Use a min-heap to keep track of the k most frequent elements
        // The heap stores Map.Entry objects, ordered by frequency (value).
        // A min-heap ensures that the smallest frequency among the top K is at the root.
        PriorityQueue<Map.Entry<Integer, Integer>> minHeap = new PriorityQueue<>(
                Comparator.comparingInt(Map.Entry::getValue)
        );

        for (Map.Entry<Integer, Integer> entry : freqMap.entrySet()) {
            minHeap.offer(entry); // Add current entry
            if (minHeap.size() > k) {
                minHeap.poll(); // Remove the entry with the smallest frequency if heap size exceeds k
            }
        }

        // Step 3: Extract the elements from the heap
        int[] result = new int[k];
        for (int i = 0; i < k; i++) {
            result[i] = minHeap.poll().getKey();
        }

        return result;
    }

    /**
     * Approach 2: Optimal Solution using HashMap + Custom Min-Heap
     *
     * This method provides the same logic as `topKFrequentHeap` but uses
     * our custom `CustomMinHeap` implementation to demonstrate heap mechanics from scratch.
     *
     * For this, we need a custom Pair class or equivalent to store element and its frequency.
     *
     * Time Complexity: O(N + M log K)
     * Space Complexity: O(M + K)
     *
     * @param nums The input integer array.
     * @param k The number of most frequent elements to return.
     * @return An array containing the k most frequent elements.
     */
    public int[] topKFrequentCustomHeap(int[] nums, int k) {
        // Custom class to store frequency and value for the custom heap
        class FrequencyPair {
            int value;
            int frequency;

            public FrequencyPair(int value, int frequency) {
                this.value = value;
                this.frequency = frequency;
            }
        }

        // Step 1: Count frequencies
        Map<Integer, Integer> freqMap = new HashMap<>();
        for (int num : nums) {
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }

        // Step 2: Use a custom min-heap to keep track of the k most frequent elements
        // Comparator ensures sorting by frequency in ascending order (for min-heap)
        CustomMinHeap<FrequencyPair> minHeap = new CustomMinHeap<>(
                Comparator.comparingInt(p -> p.frequency)
        );

        for (Map.Entry<Integer, Integer> entry : freqMap.entrySet()) {
            minHeap.add(new FrequencyPair(entry.getKey(), entry.getValue()));
            if (minHeap.size() > k) {
                minHeap.poll();
            }
        }

        // Step 3: Extract results
        int[] result = new int[k];
        for (int i = 0; i < k; i++) {
            result[i] = minHeap.poll().value;
        }

        return result;
    }


    /**
     * Approach 3: Bucket Sort (Linear Time Solution)
     *
     * This approach is generally the most optimal in terms of time complexity
     * if the maximum frequency is not excessively large (which it is bounded by N).
     *
     * Algorithm:
     * 1. **Count Frequencies:** Use a HashMap to store `(number, frequency)` pairs.
     *    Time: O(N), Space: O(M)
     *
     * 2. **Create Buckets:** Create an array of lists, where `buckets[i]` will store
     *    all numbers that have a frequency of `i`. The size of this array will be
     *    `N + 1` because the maximum possible frequency is `N`.
     *    Initialize each bucket (list) if it's null when adding.
     *    Time: O(M), Space: O(N) (worst case, if all numbers have distinct frequencies up to N)
     *
     * 3. **Populate Buckets:** Iterate through the `freqMap`. For each `(number, frequency)`
     *    pair, add `number` to `buckets[frequency]`.
     *    Time: O(M)
     *
     * 4. **Collect Top K:** Iterate through the `buckets` array from the highest
     *    possible frequency (N) down to 1. For each bucket that contains elements,
     *    add them to the result list until `k` elements have been collected.
     *    Time: O(N) (worst case, iterating through all buckets and elements)
     *
     * Overall Time Complexity: O(N)
     *   - O(N) for frequency counting.
     *   - O(N) for populating buckets (at most N elements stored across all buckets).
     *   - O(N) for collecting results (at most N elements checked).
     * Overall Space Complexity: O(N)
     *   - For frequency map and buckets array.
     *
     * @param nums The input integer array.
     * @param k The number of most frequent elements to return.
     * @return An array containing the k most frequent elements.
     */
    public int[] topKFrequentBucketSort(int[] nums, int k) {
        // Step 1: Count frequencies
        Map<Integer, Integer> freqMap = new HashMap<>();
        for (int num : nums) {
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }

        // Step 2: Create buckets for frequencies.
        // Index i represents frequency i, and the list at buckets[i] contains numbers with that frequency.
        // Max frequency can be nums.length.
        List<Integer>[] buckets = new List[nums.length + 1];

        // Step 3: Populate buckets
        for (Map.Entry<Integer, Integer> entry : freqMap.entrySet()) {
            int num = entry.getKey();
            int frequency = entry.getValue();
            if (buckets[frequency] == null) {
                buckets[frequency] = new ArrayList<>();
            }
            buckets[frequency].add(num);
        }

        // Step 4: Collect top k elements by iterating buckets from highest frequency
        List<Integer> resultList = new ArrayList<>();
        for (int i = buckets.length - 1; i >= 0 && resultList.size() < k; i--) {
            if (buckets[i] != null) {
                for (int num : buckets[i]) {
                    resultList.add(num);
                    if (resultList.size() == k) {
                        break; // Found k elements
                    }
                }
            }
        }

        return resultList.stream().mapToInt(Integer::intValue).toArray();
    }


    /**
     * Approach 4: HashMap + Sorting (for comparison)
     *
     * Algorithm:
     * 1. **Count Frequencies:** Use a HashMap to store `(number, frequency)` pairs.
     *    Time: O(N), Space: O(M)
     *
     * 2. **Convert to List and Sort:** Convert the `entrySet()` of the HashMap into an
     *    `ArrayList<Map.Entry<Integer, Integer>>`. Sort this list based on frequencies
     *    in descending order.
     *    Time: O(M log M) where M is the number of unique elements.
     *    Space: O(M)
     *
     * 3. **Extract Top K:** Take the first `k` elements from the sorted list.
     *    Time: O(K)
     *
     * Overall Time Complexity: O(N + M log M)
     *   - O(N) for frequency counting.
     *   - O(M log M) for sorting the map entries.
     *   - O(K) for extracting results.
     * Overall Space Complexity: O(M)
     *   - For frequency map and the list of entries.
     *
     * This approach is simpler but less efficient than the heap-based approach
     * if `K` is significantly smaller than `M` (the number of unique elements),
     * as it sorts all unique elements regardless of `K`.
     *
     * @param nums The input integer array.
     * @param k The number of most frequent elements to return.
     * @return An array containing the k most frequent elements.
     */
    public int[] topKFrequentSorting(int[] nums, int k) {
        // Step 1: Count frequencies
        Map<Integer, Integer> freqMap = new HashMap<>();
        for (int num : nums) {
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }

        // Step 2: Convert map entries to a list and sort by frequency in descending order
        List<Map.Entry<Integer, Integer>> entryList = new ArrayList<>(freqMap.entrySet());
        entryList.sort((e1, e2) -> e2.getValue() - e1.getValue()); // Descending order of frequency

        // Step 3: Extract the k most frequent elements
        int[] result = new int[k];
        for (int i = 0; i < k; i++) {
            result[i] = entryList.get(i).getKey();
        }

        return result;
    }
}
```

File: optimization-heap-operations-problems-complete/src/problems/TopKFrequentElements.ts
```typescript
import { Heap } from '../data-structures/Heap';
import { defaultCompare } from '../utils/comparator';

/**
 * Finds the K most frequent elements in an array using a frequency map and a Min-Heap.
 *
 * Algorithm:
 * 1. Count frequencies: Iterate through `nums` and store element frequencies in a `Map<number, number>`.
 * 2. Build a Min-Heap:
 *    a. Create a Min-Heap that stores objects of type `{ element: number, frequency: number }`.
 *    b. The comparison function for the heap should prioritize elements with *lower* frequency.
 *    c. Iterate through the `frequencyMap`:
 *       i. Insert each `{ element, frequency }` pair into the heap.
 *       ii. If the heap's size exceeds `k`, remove the element with the smallest frequency (root).
 * 3. Extract results: After processing all elements, the heap will contain the `k` most frequent elements.
 *    Extract them one by one to form the result array.
 *
 * Why this works:
 * A Min-Heap of size `k` storing `[frequency, element]` pairs will always have the *least frequent* of the top `k` elements at its root.
 * When a new element's frequency is encountered:
 * - If the heap size is less than `k`, add it.
 * - If the heap size is `k`, and the new element's frequency is *greater* than the root's frequency,
 *   it means the new element is more frequent than at least one of the current top `k`.
 *   So, we remove the least frequent element from the heap (root) and add the new element.
 * - If the new element's frequency is *less than or equal to* the root's frequency, it's not among the top `k`, so we do nothing.
 *
 * Time Complexity: O(N + M log K)
 *   - O(N) for building the frequency map (where N is the length of `nums`).
 *   - O(M log K) for heap operations, where M is the number of unique elements. Each of M unique elements might be inserted into the heap. Heap operations are O(log K). In the worst case, M can be N. So, O(N log K).
 * Space Complexity: O(M + K)
 *   - O(M) for the frequency map (stores M unique elements).
 *   - O(K) for the heap (stores K elements).
 *   - Total: O(N) in worst case (all elements unique).
 *
 * @param nums The input array of integers.
 * @param k The number of most frequent elements to return.
 * @returns An array containing the k most frequent elements.
 */
export function topKFrequent(nums: number[], k: number): number[] {
    if (!nums || nums.length === 0 || k <= 0) {
        return [];
    }

    // Step 1: Count frequencies of each number
    const frequencyMap = new Map<number, number>();
    for (const num of nums) {
        frequencyMap.set(num, (frequencyMap.get(num) || 0) + 1);
    }

    // Step 2: Use a Min-Heap to keep track of the k most frequent elements
    // The heap will store objects { element: number, frequency: number }
    // We want a Min-Heap based on frequency, so the smallest frequency will be at the root.
    const freqComparator = (a: { element: number, frequency: number }, b: { element: number, frequency: number }) => {
        return defaultCompare(a.frequency, b.frequency);
    };
    const minHeap = Heap.createMinHeap<{ element: number, frequency: number }>(freqComparator);

    for (const [element, frequency] of frequencyMap.entries()) {
        minHeap.insert({ element, frequency });
        // If the heap size exceeds k, remove the element with the smallest frequency (the root)
        if (minHeap.size() > k) {
            minHeap.extract();
        }
    }

    // Step 3: Extract elements from the heap
    const result: number[] = [];
    while (!minHeap.isEmpty()) {
        result.push(minHeap.extract()!.element);
    }

    // The elements are extracted in ascending order of frequency (due to min-heap),
    // but the problem doesn't specify order, so this is fine.
    // If order by frequency (descending) was needed, could reverse or use a Max-Heap and extract.
    return result;
}

/**
 * Alternative approach: Finds the K most frequent elements by sorting.
 *
 * Algorithm:
 * 1. Count frequencies: Iterate through `nums` and store element frequencies in a `Map<number, number>`.
 * 2. Convert map to array: Transform the frequency map into an array of `[element, frequency]` pairs.
 * 3. Sort by frequency: Sort this array in descending order of frequency.
 * 4. Extract top K: Take the first `k` elements from the sorted array.
 *
 * Time Complexity: O(N + M log M)
 *   - O(N) for building the frequency map.
 *   - O(M) for converting map to array.
 *   - O(M log M) for sorting the array of M unique elements (M can be up to N).
 * Space Complexity: O(N)
 *   - O(N) for the frequency map (worst case all elements unique).
 *   - O(N) for the array of `[element, frequency]` pairs.
 *
 * This approach is simpler but can be less efficient than the heap approach if M is large
 * (e.g., all elements are unique, M=N), making the sort O(N log N) while the heap is O(N log K).
 * If K is close to N, then both are O(N log N). If K is small, heap is better.
 *
 * @param nums The input array of integers.
 * @param k The number of most frequent elements to return.
 * @returns An array containing the k most frequent elements.
 */
export function topKFrequentSort(nums: number[], k: number): number[] {
    if (!nums || nums.length === 0 || k <= 0) {
        return [];
    }

    // Step 1: Count frequencies
    const frequencyMap = new Map<number, number>();
    for (const num of nums) {
        frequencyMap.set(num, (frequencyMap.get(num) || 0) + 1);
    }

    // Step 2 & 3: Convert to array of [element, frequency] pairs and sort by frequency (descending)
    const sortedFrequencies = Array.from(frequencyMap.entries()).sort((a, b) => b[1] - a[1]);

    // Step 4: Extract the top k elements
    const result: number[] = [];
    for (let i = 0; i < k && i < sortedFrequencies.length; i++) {
        result.push(sortedFrequencies[i][0]);
    }

    return result;
}
```